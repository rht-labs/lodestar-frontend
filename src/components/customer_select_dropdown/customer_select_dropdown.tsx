import React, { useState, useEffect } from 'react';
import { Select, SelectVariant, SelectOption } from '@patternfly/react-core';
import { useEngagements } from '../../context/engagement_context/engagement_hook';
export interface CustomerSelectDropdownProps {
  onSelect: (selectedOption: string) => void;
  selectedValue?: string;
  placeholder?: string;
}
export interface HasOptions {
  options?: string[];
}
export function _CustomerSelectDropdown(
  props: CustomerSelectDropdownProps & HasOptions
) {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const handleSelect = (_, selection) => {
    props.onSelect(selection);
    setIsOpen(false);
  };
  return (
    <Select
      isOpen={isOpen}
      onToggle={() => setIsOpen(!isOpen)}
      onSelect={handleSelect}
      maxHeight={400}
      isCreatable={true}
      selections={props.selectedValue}
      variant={SelectVariant.typeahead}
      placeholderText={props.placeholder}
      onClear={() => props.onSelect(undefined)}
      toggleId="customer_dropdown"
    >
      {props.options
        ?.concat([props.selectedValue])
        .filter(v => !!v)
        .map(value => (
          <SelectOption data-testid={value} value={value} key={value} />
        ))}
    </Select>
  );
}

export function CustomerSelectDropdown(props: CustomerSelectDropdownProps) {
  const { engagements, getEngagements } = useEngagements();
  const customerNames = Array.from(
    new Set(
      engagements?.filter(e => !!e.customer_name).map(e => e.customer_name) ??
        []
    )
  );
  const CUSTOMER_DROPDOWN = _CustomerSelectDropdown;
  useEffect(() => {
    if (engagements === undefined) {
      getEngagements();
    }
  }, [engagements, getEngagements]);
  return (
    <CUSTOMER_DROPDOWN
      onSelect={props.onSelect}
      options={customerNames}
      placeholder={props.placeholder}
      selectedValue={props.selectedValue}
    />
  );
}
