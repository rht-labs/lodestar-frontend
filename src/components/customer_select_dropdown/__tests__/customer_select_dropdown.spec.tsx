import React from 'react';
import { render, fireEvent, queryByAttribute } from '@testing-library/react';
import { CustomerSelect } from '../customer_select_dropdown';

describe('Customer Select Dropdown', () => {
  let customers;
  beforeEach(() => {
    customers = ['foo', 'bar', 'baz'];
  });
  test('matches snapshot', () => {
    expect(
      render(
        <CustomerSelect
          onSelect={() => {}}
          options={customers}
          placeholder="Foobar"
        />
      )
    ).toMatchSnapshot();
  });

  test('shows existing customers', async () => {
    const wrapper = render(
      <CustomerSelect
        onSelect={() => {}}
        options={customers}
        placeholder="Foobar"
      />
    );
    const dropdownButton = await wrapper.findByRole('button');
    await fireEvent.click(dropdownButton);
    const fooCustomer = await wrapper.findByTestId('foo');
    expect(fooCustomer).toBeDefined();
  });

  test('can select existing customer', async () => {
    const selectCallback = jest.fn();
    const wrapper = render(
      <CustomerSelect
        onSelect={selectCallback}
        options={customers}
        placeholder="Foobar"
      />
    );
    const dropdownButton = await wrapper.findByRole('button');
    await fireEvent.click(dropdownButton);
    const fooCustomer = await wrapper.findByTestId('foo');
    await fireEvent.click(fooCustomer);
    expect(selectCallback).toHaveBeenCalled();
  });

  test('can create a new customer', async () => {
    const selectCallback = jest.fn();

    const wrapper = render(
      <CustomerSelect
        onSelect={selectCallback}
        options={customers}
        placeholder="Foobar"
      />
    );
    const dropdownButton = await wrapper.findByRole('button');

    const selectTextInput = await wrapper.findByRole('textbox');
    await fireEvent.click(dropdownButton);
    await fireEvent.change(selectTextInput, { target: { value: 'bot' } });
    const newItem = await wrapper.findByRole('option');
    await fireEvent.click(newItem);
    expect(selectCallback).toBeCalledWith('bot');
  });
});
