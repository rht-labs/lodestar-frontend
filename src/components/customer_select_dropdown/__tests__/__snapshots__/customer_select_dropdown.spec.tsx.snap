// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer Select Dropdown matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pf-c-select"
        data-ouia-component-id="0"
        data-ouia-component-type="PF4/Select"
        data-ouia-safe="true"
      >
        <div
          class="pf-c-select__toggle pf-m-typeahead"
        >
          <div
            class="pf-c-select__toggle-wrapper"
          >
            <input
              aria-label=""
              autocomplete="off"
              class="pf-c-form-control pf-c-select__toggle-typeahead"
              id="customer_dropdown-select-typeahead"
              placeholder="Foobar"
              type="text"
              value=""
            />
          </div>
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-label="Options menu"
            aria-labelledby=" customer_dropdown"
            class="pf-c-button pf-c-select__toggle-button pf-m-plain"
            id="customer_dropdown"
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="pf-c-select__toggle-arrow"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                transform=""
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pf-c-select"
      data-ouia-component-id="0"
      data-ouia-component-type="PF4/Select"
      data-ouia-safe="true"
    >
      <div
        class="pf-c-select__toggle pf-m-typeahead"
      >
        <div
          class="pf-c-select__toggle-wrapper"
        >
          <input
            aria-label=""
            autocomplete="off"
            class="pf-c-form-control pf-c-select__toggle-typeahead"
            id="customer_dropdown-select-typeahead"
            placeholder="Foobar"
            type="text"
            value=""
          />
        </div>
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-label="Options menu"
          aria-labelledby=" customer_dropdown"
          class="pf-c-button pf-c-select__toggle-button pf-m-plain"
          id="customer_dropdown"
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="pf-c-select__toggle-arrow"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              transform=""
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
